File: test/language/sql_constraints/cpp/SqlConstraintsTest.cpp
#include <cstdio>
#include <vector>

#include "gtest/gtest.h"

#include "zserio/SqliteException.h"

#include "sql_constraints/TestDb.h"
#include "sql_constraints/ConstraintsTable.h"
#include "sql_constraints/ConstraintsConstant.h"
#include "sql_constraints/constraints/ImportedConstant.h"
#include "sql_constraints/constraints/ImportedEnum.h"

namespace sql_constraints
{

class SqlConstraintsTest : public ::testing::Test
{
public:
    SqlConstraintsTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new TestDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~SqlConstraintsTest()
    {
        delete m_database;
    }

protected:
    void fillRow(ConstraintsTable::Row& row)
    {
        row.setWithoutSql(1);
        row.setSqlNotNull(1);
        row.setSqlDefaultNull(1);
        row.setSqlNull(1);
        row.setSqlCheckConstant(1);
        row.setSqlCheckImportedConstant(1);
        row.setSqlCheckEnum(ConstraintsEnum::VALUE1);
        row.setSqlCheckImportedEnum(constraints::ImportedEnum::ONE);
        row.setSqlCheckUnicodeEscape(UNICODE_ESCAPE_CONST);
        row.setSqlCheckHexEscape(HEX_ESCAPE_CONST);
        row.setSqlCheckOctalEscape(OCTAL_ESCAPE_CONST);
    }

    static const char DB_FILE_NAME[];

    static const uint8_t UNICODE_ESCAPE_CONST;
    static const uint8_t HEX_ESCAPE_CONST;
    static const uint8_t OCTAL_ESCAPE_CONST;

    static const uint8_t WRONG_UNICODE_ESCAPE_CONST;
    static const uint8_t WRONG_HEX_ESCAPE_CONST;
    static const uint8_t WRONG_OCTAL_ESCAPE_CONST;

    TestDb* m_database;
};

const char SqlConstraintsTest::DB_FILE_NAME[] = "sql_constraints_test.sqlite";

const uint8_t SqlConstraintsTest::UNICODE_ESCAPE_CONST = 1;
const uint8_t SqlConstraintsTest::HEX_ESCAPE_CONST = 2;
const uint8_t SqlConstraintsTest::OCTAL_ESCAPE_CONST = 3;

const uint8_t SqlConstraintsTest::WRONG_UNICODE_ESCAPE_CONST = 0;
const uint8_t SqlConstraintsTest::WRONG_HEX_ESCAPE_CONST = 0;
const uint8_t SqlConstraintsTest::WRONG_OCTAL_ESCAPE_CONST = 0;

TEST_F(SqlConstraintsTest, withoutSql)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.resetWithoutSql();
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlNotNull)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.resetSqlNotNull();
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlDefaultNull)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.resetSqlDefaultNull();
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_NO_THROW(constraintsTable.write(rows));
}

TEST_F(SqlConstraintsTest, sqlNull)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.resetSqlNull();
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_NO_THROW(constraintsTable.write(rows));
}

TEST_F(SqlConstraintsTest, sqlCheckConstant)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckConstant(ConstraintsConstant);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlCheckImportedConstant)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckImportedConstant(constraints::ImportedConstant);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlCheckEnum)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckEnum(ConstraintsEnum::VALUE2);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlCheckImportedEnum)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckImportedEnum(constraints::ImportedEnum::TWO);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlCheckUnicodeEscape)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckUnicodeEscape(WRONG_UNICODE_ESCAPE_CONST);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlCheckHexEscape)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckHexEscape(WRONG_HEX_ESCAPE_CONST);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

TEST_F(SqlConstraintsTest, sqlCheckOctalEscape)
{
    ConstraintsTable& constraintsTable = m_database->getConstraintsTable();
    ConstraintsTable::Row row;
    fillRow(row);
    row.setSqlCheckOctalEscape(WRONG_OCTAL_ESCAPE_CONST);
    std::vector<ConstraintsTable::Row> rows;
    rows.push_back(row);
    ASSERT_THROW(constraintsTable.write(rows), zserio::SqliteException);
}

} // namespace sql_constraints

File: test/language/sql_constraints/python/SqlConstraintsTest.py
import unittest
import os
import apsw

from testutils import getZserioApi, getApiDir

class SqlConstraintsTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "sql_constraints.zs")
        cls._fileName = os.path.join(getApiDir(os.path.dirname(__file__)), "sql_constraints_test.sqlite")

    def setUp(self):
        if os.path.exists(self._fileName):
            os.remove(self._fileName)
        self._database = self.api.TestDb.fromFile(self._fileName)
        self._database.createSchema()
        self._constraintsTable = self._database.getConstraintsTable()

    def tearDown(self):
        self._database.close()

    def testWithoutSql(self):
        rowDict = self._createRowDict()
        rowDict['withoutSql'] = None
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("NOT NULL constraint failed: constraintsTable.withoutSql" in str(context.exception))

    def testSqlNotNull(self):
        rowDict = self._createRowDict()
        rowDict['sqlNotNull'] = None
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("NOT NULL constraint failed: constraintsTable.sqlNotNull" in str(context.exception))

    def testSqlDefaultNull(self):
        rowDict = self._createRowDict()
        rowDict['sqlDefaultNull'] = None
        self._writeRowDict(rowDict)

    def testSqlNull(self):
        rowDict = self._createRowDict()
        rowDict['sqlNull'] = None
        self._writeRowDict(rowDict)

    def testSqlCheckConstant(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckConstant'] = self.api.ConstraintsConstant
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckImportedConstant(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckImportedConstant'] = self.api.constraints.ImportedConstant
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckEnum(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckEnum'] = self.api.ConstraintsEnum.VALUE2
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckImportedEnum(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckImportedEnum'] = self.api.constraints.ImportedEnum.TWO
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckBitmask(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckBitmask'] = self.api.ConstraintsBitmask.Values.MASK2
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckImportedBitmask(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckImportedBitmask'] = self.api.constraints.ImportedBitmask.Values.MASK2
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckUnicodeEscape(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckUnicodeEscape'] = self.WRONG_UNICODE_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckHexEscape(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckHexEscape'] = self.WRONG_HEX_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def testSqlCheckOctalEscape(self):
        rowDict = self._createRowDict()
        rowDict['sqlCheckOctalEscape'] = self.WRONG_OCTAL_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
            self.assertTrue("CHECK constraint failed: constraintsTable" in str(context.exception))

    def _createRowDict(self):
        return {
            'primaryKey' : 1,
            'withoutSql' : 1,
            'sqlNotNull' : 1,
            'sqlDefaultNull' : 1,
            'sqlNull' : 1,
            'sqlCheckConstant' : 1,
            'sqlCheckImportedConstant' : 1,
            'sqlCheckEnum' : self.api.ConstraintsEnum.VALUE1,
            'sqlCheckImportedEnum' : self.api.constraints.ImportedEnum.ONE,
            'sqlCheckBitmask' : self.api.ConstraintsBitmask.Values.MASK1,
            'sqlCheckImportedBitmask' : self.api.constraints.ImportedBitmask.Values.MASK1,
            'sqlCheckUnicodeEscape' : self.UNICODE_ESCAPE_CONST,
            'sqlCheckHexEscape' : self.HEX_ESCAPE_CONST,
            'sqlCheckOctalEscape' : self.OCTAL_ESCAPE_CONST
        }

    def _writeRowDict(self, rowDict):
        row = [(
            rowDict['primaryKey'],
            rowDict['withoutSql'],
            rowDict['sqlNotNull'],
            rowDict['sqlDefaultNull'],
            rowDict['sqlNull'],
            rowDict['sqlCheckConstant'],
            rowDict['sqlCheckImportedConstant'],
            rowDict['sqlCheckEnum'],
            rowDict['sqlCheckImportedEnum'],
            rowDict['sqlCheckBitmask'],
            rowDict['sqlCheckImportedBitmask'],
            rowDict['sqlCheckUnicodeEscape'],
            rowDict['sqlCheckHexEscape'],
            rowDict['sqlCheckOctalEscape']
        )]
        self._constraintsTable.write(row)

    UNICODE_ESCAPE_CONST = 1
    HEX_ESCAPE_CONST = 2
    OCTAL_ESCAPE_CONST = 3

    WRONG_UNICODE_ESCAPE_CONST = 0
    WRONG_HEX_ESCAPE_CONST = 0
    WRONG_OCTAL_ESCAPE_CONST = 0
